designs/component-upgrades.md

```diff
@@ -0,0 +1,149 @@
+* Upgrade cli
+* Upgrades to customer defined component versions
+
+**Future scope**
+
+*  Version changes report
+* “Notification” for new available versions
+* Automatic/scheduled upgrades
+
+## Overview of solution
+
+**Upgrade all eks-a components to the versions declared in the latest bundles manifest when running the `upgrade cluster` command.**
+
+## Solution details
+
+Since we already have:
+
+* An interface that customers use to upgrade clusters (`upgrade cluster` command)
+* An declarative list of all eks-a components at their latest available versions that have been tested together (bundles manifest)
+
+It seems like the simplest solution is to take advantage of both.
+
+This solution fits very well with both, the goal of a customer experience as simple as possible and the concept of the versions bundle (a group of versioned components that work well together):
+
+* We don’t introduce more concepts/commands. Customers just have to use the same command they are already used to. The syntax stays the same and there is no need to define what needs to be upgraded and to what version, the cli takes care of this transparently.
+* Components bundled together are always upgraded together. This reduces our testing matrix considerably, since there is no room for version miss-alignments.
+
+The cli needs the ability to detect both a component version changes as well as new builds of the same versioned component (eg. new container build due to a patch in a dependency like AL2). The process to upgrade some components could be different for these two scenarios.
+
+### Order of execution
+
+Components should be **upgraded** **before the new Kubernetes version is rolled out** (if needed). It’s more common for an old component to have problems with a new Kubernetes version than vice-versa. There might be case by case exceptions, right now there is not, at least with the current set of core components.
+
+This also facilitates CAPI api version upgrades, since updating its components (CRDs and controllers) will also update the current objects in the cluster (through conversion webhooks). This allows us to only maintain only one api version of provider templates, since these are only reapplied when rolling out the new Kubernetes version.
+
+### How to detect new builds for components?
+
+All the CAPI providers objects in the bundles manifest already contain a `version` field. Unfortunately it only reflects the original upstream version that component was built from. We should add the build metadata information to that semver, like `v0.4.2+build-2`. This will allow us to detect new builds as well as preserve the original upstream version, which is important in order to verify api version changes. We might need to trim the build metadata info when interacting with `clusterctl`, I haven’t tested if it’s supported yet.
+
+We will need to add such field to all the other components (Flux, eks-a controllers and crds, cilium...) that don’t have it yet.
```

Comment:
We also need to cross-reference with what is already installed, meaning having it in our Cluster status for example on what bundle is currently being consumed for the cluster. Sounds like we need to pick this issue up for that then? https://github.com/aws/eks-anywhere/issues/165

Reply:
That's a good point
However, it's not completely necessary. The current controller is capable of retrieving the installed `VersionsBundle`, so the cli should be able to do the same (following the same naming assumptions)

That said, I think using the status is a way better solution. So if we have bandwidth, it would be ideal to make it part of this project. Maybe we can implement a centralized logic to get the "current `VersionBundle`" and use it from the cli and controller. And whenever we pick that ticket, just update that place, whether is a part of this project or later.

Reply:
Added to project task list